File: client/src/validations/contactSchema.ts
import { z } from 'zod';

export const contactSchema = z.object({
  full_name: z
    .string()
    .min(3, 'Name must be at least 3 characters')
    .max(100, 'Name must be less than 100 characters')
    .regex(/^[a-zA-Z\s'-]+$/, 'Name can only contain letters, spaces, hyphens, and apostrophes'),
  phone: z
    .string()
    .min(10, 'Please enter a valid phone number')
    .regex(/^[+]?[\d\s\-()]{10,15}$/, 'Please enter a valid phone number'),
  email: z
    .string()
    .email('Please enter a valid email address')
    .max(255, 'Email address is too long'),
  introduction: z
    .string()
    .min(20, 'Introduction must be at least 20 characters')
    .max(1000, 'Introduction must be less than 1000 characters')
    .trim(),
  purpose: z
    .string()
    .min(1, 'Please select a purpose')
    .refine((val: string) => ['Job Opportunity', 'Freelance Project', 'Collaboration', 'Just Saying Hi'].includes(val), 'Invalid purpose selected'),
  company: z
    .string()
    .min(2, 'Company name must be at least 2 characters')
    .max(100, 'Company name must be less than 100 characters')
    .trim(),
  budget: z
    .string()
    .refine(
      (val: string) => !val || ['Less than $500', '$500–$2,000', '$2,000–$5,000', '$5,000+'].includes(val),
      'Invalid budget range selected'
    )
    .optional()
    .or(z.literal('')),
  source: z
    .string()
    .refine(
      (val: string) => !val || ['LinkedIn', 'GitHub', 'Google Search', 'Referral', 'Other'].includes(val),
      'Invalid source selected'
    )
    .optional()
    .or(z.literal('')),
  pdf_filename: z
    .string()
    .max(100, 'PDF filename must be less than 100 characters')
    .refine(
      (val: string) => !val || /^[a-zA-Z0-9_\s-]+$/.test(val),
      'PDF filename can only contain letters, numbers, spaces, hyphens, and underscores'
    )
    .optional()
    .or(z.literal('')),
}).refine(
  data => data.purpose !== 'Freelance Project' || data.budget,
  {
    message: 'Budget range is required for Freelance Projects',
    path: ['budget'],
  }
);

export type ContactFormValues = z.infer<typeof contactSchema>;

File: server/app/schemas/contact_schema.py
from marshmallow import Schema, fields, validate, ValidationError, pre_load, post_load
import re


def validate_filename(value):
    """Validate PDF filename - only alphanumeric, spaces, hyphens, underscores"""
    if value and not re.match(r'^[a-zA-Z0-9_\s-]+$', value):
        raise ValidationError('PDF filename can only contain letters, numbers, spaces, hyphens, and underscores')


def validate_name(value):
    """Validate full name - only letters, spaces, hyphens, apostrophes"""
    if not re.match(r"^[a-zA-Z\s'-]+$", value):
        raise ValidationError('Name can only contain letters, spaces, hyphens, and apostrophes')


class ContactSchema(Schema):
    full_name = fields.Str(
        required=True,
        validate=[validate.Length(min=3, max=100), validate_name]
    )
    phone = fields.Str(required=True, validate=validate.Length(min=10, max=15))
    email = fields.Email(required=True, validate=validate.Length(max=255))
    introduction = fields.Str(
        required=True,
        validate=validate.Length(min=20, max=1000)
    )
    purpose = fields.Str(
        required=True,
        validate=validate.OneOf(['Job Opportunity', 'Freelance Project', 'Collaboration', 'Just Saying Hi'])
    )
    company = fields.Str(required=True, validate=validate.Length(min=2, max=100))
    budget = fields.Str(
        required=False,
        validate=validate.OneOf(['Less than $500', '$500–$2,000', '$2,000–$5,000', '$5,000–$10,000', '$10,000+', 'Not Specified']),
        allow_none=True
    )
    source = fields.Str(
        required=False,
        validate=validate.OneOf(['LinkedIn', 'GitHub', 'Portfolio Website', 'Google Search', 'Referral', 'Other']),
        allow_none=True
    )
    pdf_filename = fields.Str(
        required=False,
        validate=[validate.Length(max=100), validate_filename],
        allow_none=True,
        load_default=''
    )

    @pre_load
    def preprocess(self, data, **kwargs):
        """Trim whitespace from string fields"""
        for key in ['full_name', 'company', 'introduction']:
            if key in data and isinstance(data[key], str):
                data[key] = data[key].strip()
        return data


class LoginSchema(Schema):
    username = fields.Str(required=True, validate=validate.Length(min=3))
    password = fields.Str(required=True, validate=validate.Length(min=6))


contact_schema = ContactSchema()
login_schema = LoginSchema()
